x/gov/keeper/collections_pagination_gov.go

```diff
@@ -0,0 +1,330 @@
+					if err != nil {
+						return nil, nil, err
+					}
+					results = append(results, transformed)
+					count++
+				}
+			}
+		// second case, we found all the objects specified within the limit
+		case count == limit:
+			key, err := iterator.Key()
+			if err != nil {
+				if strings.Contains(err.Error(), "no concrete type registered for type URL") {
+					// URL /osmosis.concentratedliquidity.v1beta1.CreateConcentratedLiquidityPoolsProposal
+					continue
+				}
+				return nil, nil, err
+			}
+			nextKey, err = encodeCollKey[K, V](coll, key)
+			if err != nil {
+				return nil, nil, err
+			}
+			// if count total was not specified, we return the next key only
+			if !countTotal {
+				return results, &query.PageResponse{
+					NextKey: nextKey,
+				}, nil
+			}
+			// otherwise we fallthrough the third case
+			fallthrough
+		// this is the case in which we found all the required results
+		// but we need to count how many possible results exist in total.
+		// so we keep increasing the count until the iterator is fully consumed.
+		case count > limit:
+			if predicateFunc == nil {
+				count++
+
+				// if predicate function is defined we check if the result matches the filtering criteria
+			} else {
+				kv, err := iterator.KeyValue()
+				if err != nil {
+					if strings.Contains(err.Error(), "no concrete type registered for type URL") {
+						// URL /osmosis.concentratedliquidity.v1beta1.CreateConcentratedLiquidityPoolsProposal
+						continue
+					}
+					return nil, nil, err
+				}
+
+				include, err := predicateFunc(kv.Key, kv.Value)
+				if err != nil {
+					return nil, nil, err
+				}
+				if include {
+					count++
+				}
+			}
+		}
+	}
+
+	resp := &query.PageResponse{
+		NextKey: nextKey,
+	}
+
+	if countTotal {
+		resp.Total = count + offset
+	}
+	return results, resp, nil
+}
+
+// collFilteredPaginateByKey paginates a collection when a starting key
+// is provided in the PageRequest. Predicate is applied only if not nil.
+func collFilteredPaginateByKey[K, V any, C collq.Collection[K, V], T any](
+	ctx context.Context,
+	coll C,
+	prefix []byte,
+	key []byte,
+	reverse bool,
+	limit uint64,
+	predicateFunc func(key K, value V) (bool, error),
+	transformFunc func(key K, value V) (transformed T, err error),
+) (results []T, pageRes *query.PageResponse, err error) {
+	iterator, err := getCollIter[K, V](ctx, coll, prefix, key, reverse)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer iterator.Close()
+
+	var (
+		count   uint64
+		nextKey []byte
+	)
+
+	for ; iterator.Valid(); iterator.Next() {
+		// if we reached the specified limit
+		// then we get the next key, and we exit the iteration.
+		if count == limit {
+			concreteKey, err := iterator.Key()
+			if err != nil {
+				return nil, nil, err
+			}
+
+			nextKey, err = encodeCollKey[K, V](coll, concreteKey)
+			if err != nil {
+				return nil, nil, err
+			}
+			break
+		}
+
+		kv, err := iterator.KeyValue()
+		if err != nil {
```

Comment:
Are we missing the fix here as well? This is called when the `key` param is provided for pagination afaik

Reply:
Good shout added